src/System.Net.Mime/src/System/Net/Mail/SmtpLoginAuthenticationModule.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.IO;
+using System.Net;
+using System.Security.Permissions;
+using System.Security.Authentication.ExtendedProtection;
+
+namespace System.Net.Mail
+{
+    internal class SmtpLoginAuthenticationModule : ISmtpAuthenticationModule
+    {
+        private Hashtable _sessions = new Hashtable();
+
+        internal SmtpLoginAuthenticationModule()
+        {
+        }
+
+        #region ISmtpAuthenticationModule Members
+
+        // Security this method will access NetworkCredential properties that demand UnmanagedCode and Environment Permission
+        [EnvironmentPermission(SecurityAction.Assert, Unrestricted = true)]
+        [SecurityPermission(SecurityAction.Assert, Flags = SecurityPermissionFlag.UnmanagedCode)]
+        public Authorization Authenticate(string challenge, NetworkCredential credential, object sessionCookie, string spn, ChannelBinding channelBindingToken)
+        {
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Enter(NetEventSource.ComponentType.Web, this, "Authenticate", null);
+            try
+            {
+                lock (_sessions)
+                {
+                    NetworkCredential cachedCredential = _sessions[sessionCookie] as NetworkCredential;
```

Comment:
If/when `_sessions` changes to `Dictionary`, `TryGetValue` will need to be used here.